repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilterLoader.scala

```diff
@@ -0,0 +1,321 @@
+      //Get the specific class to load from the application context
+      val filterClass: Class[_] = filterClassLoader.loadClass(filterClassName)
+      val newFilterInstance: Filter = try {
+        filterContext.getBean(filterClass).asInstanceOf[Filter]
+      } catch {
+        case e: NoSuchBeanDefinitionException =>
+          logger.debug("Could not load the filter {} using Spring. Will try to manually load the class instead.", filterClassName, e)
+          //Spring didn't load the filter as a bean, try manually creating a new instance of the class
+          val rtnFilterInstance = filterClass.newInstance.asInstanceOf[Filter]
+          //Add the instance to the application context using its full class name
+          filterContext.getBeanFactory.registerSingleton(rtnFilterInstance.getClass.getName, rtnFilterInstance)
+          rtnFilterInstance
+      }
+      newFilterInstance.init(new FilterConfigWrapper(servletContext, filterType, filter.getConfiguration))
+      logger.info("Filter Instance: {} successfully created", newFilterInstance)
+      FilterContext(newFilterInstance, filter.getName, filter.getFilterCriterion.evaluate, filterContext, filter)
+    } catch {
+      case e: ClassNotFoundException =>
+        throw new FilterInitializationException(s"Requested filter, $filterClassName does not exist in any loaded artifacts", e)
+      case e: ServletException =>
+        val message = s"Failed to initialize filter $filterClassName"
+        logger.error(message)
+        throw new FilterInitializationException(message, e)
+      case e: ClassCastException =>
+        throw new FilterInitializationException(s"Requested filter, $filterClassName is not of type javax.servlet.Filter", e)
+      case e@(_: InstantiationException | _: IllegalAccessException) =>
+        throw new FilterInitializationException(s"Requested filter, $filterClassName is not an annotated Component nor does it have a public zero-argument constructor.", e)
+    }
+  }
+
+  private def getUniqueContextName(filterInfo: FilterConfig): String = {
+    val sb = new StringBuilder
+    if (filterInfo.getId != null) sb.append(filterInfo.getId).append("-")
+    sb.append(filterInfo.getName).append("-")
+    sb.append(UUID.randomUUID.toString)
+    sb.toString
+  }
+}
+
+object ReposeFilterLoader {
+
+  case class FilterContext(filter: Filter, filterName: String, shouldRun: HttpServletRequestWrapper => Boolean, appContext: AbstractApplicationContext, filterConfig: FilterConfig)
+
+  class FilterContextList(val registrar: FilterContextRegistrar, val filterContexts: List[FilterContext], val bypassUriRegex: Option[String]) extends AutoCloseable {
+    override def close(): Unit = registrar.release(this)
+  }
+
+  class FilterContextRegistrar(val filterContexts: List[FilterContext], val bypassUriRegex: Option[String]) extends AutoCloseable {
+    val pool = new util.ArrayList[FilterContextList]()
```

Comment:
Does `inUse` sound better?

Reply:
`inUse` sounds good to me.

Why use a mutable `ListBuffer`, though? Best practice seems to be to prefer mutable references to immutable data structures over immutable references to mutable data structures. Especially considering that `pool` is only used within this class. Also, do you actually need to maintain order? It does not appear so, in which case a `Set` might  be more appropriate. An immutable `HashSet` even gives you effectively constant time adds and removes. So I would think you could make it:
`private var inUse = Set.empty[FilterContextList]`

Then to add:
`inUse += newListReference`
And to remove:
`inUse -= oldListReference`

This approach has the added benefit of not requiring synchronization since the collection is immutable and (as it keeps coming up) assignment operations are atomic.

Reply:
I updated the variable name to `inUse`.

I don't have strong feelings either way about whether the reference or the data structure is the mutable entity or what the data structure is, so it is updated to a `private var inUse = Set.empty[FilterContextList]`.

However, I'm not convinced about the `synchronized` blocks no longer being necessary.
To me, they become more necessary with the `inUse` data structure being referenced that is changing.
Otherwise a new `FilterContextList` **could** be created with one `inUse` reference on the first thread.
Then before the `inUse` reference is updated/replaced, another new `FilterContextList` is created with the same `inUse` reference on a second thread.
Then the `inUse` reference is updated/replaced by the first thread adding the first `FilterContextList`.
When the `inUse` reference is updated/replaced by the second thread the first `FilterContextList` is not referenced, only the second `FilterContextList`.
This would result in an orphaned unknown `FilterContextList` that could have the filter/contexts destroyed/closed.

To further entangle this, without a `synchronized` block in the `checkShutdown`, a `close` could result in the same thing only the filter/contexts are destroyed/closed while a `FilterContextList` is being created.

Reply:
Ah, you're right. If either `+=` or `-=` are not atomic, since they are more than just assignments, there is an issue if synchronization is not used. And your second case is also an issue.

If `-=` is atomic, you could drop the synchronization on removal, but that makes the code less obvious so I wouldn't recommend it.

That is to say, this looks fine as it is now.